README.md
# Overview

### What Carpenter is

Carpenter is a set of tools designed to help you play with Dynamo tables on your local machine. 

### What Carpenter is NOT

Carpenter is NOT designed to be used to manage your infrastructure against the real AWS DynamoDB service. For that use infrastructure as code tools like terraform.

## Carpenter functions

### createTable
Carpenter creates a table. If a name is not provided through the *tableName* parameter Carpenter will generate a uuid name. Is it also possible to create secondary indexes with the table via *gsiNumber* and *lsiNumber* parameter. For each global secondary index a couple of additional parameters (GSI<_number_>PK and GSI<_number_>SK) will be created. For local secondary indexes only one additional parameter will be created (LSI<_number_>SK);

### deleteTable
Carpenter deletes a table, a table name must be provided.

### listTables
Carpenter provides a list of the tables available in Dynamo. 

### dynamoUp
Carpenter starts *instructure/dynamo-local-admin* Dynamo docker image which is based on amazon/dynamodb-local image and provides a nice GUI accessible at http://localhost:<_dynamoPort_>. A *dynamoPort* parameter can also be provided. 

### dynamoDown
Carpenter stops Dynamo.

### scanTable
Carpenter scan a table and provide a count of the items found in the table. The scan is recursive, a table name must be provided. If you also want to return a list of all the documents provide *--returnDocuments* set to true. 

### truncate
Carpenter truncates a table. A table name must be provided.

## Usage
Clone carpenter

```
git clone https://github.com/pikselpalette/carpenter.git
```

Use carpenter:

```
node bin/index --help
```

Example of a create table:

```
node bin/index --function createTable --tableName identity_local --gsiNumber 15 --lsiNumber 5
```

changelog.md
## unreleased

## v0.1.0 - 2021-03-24

  * initial project created


lib/functions.js
const AWS = require('aws-sdk');
const _ = require('lodash');

const awsConfig = {
  region: 'local',
  endpoint: 'http://localhost:8000',
  accessKeyId: 'unittest',
  secretAccessKey: 'letmein',
  maxRetries: 5
};

const dynamoOptions = { apiVersion: '2012-08-10' };

const getDynamoClient = (dynamoPort) => {
  if (dynamoPort) {
    awsConfig.endpoint = `http://localhost:${dynamoPort}`;
  }
  AWS.config.update(awsConfig);
  AWS.config.setPromisesDependency();
  return new AWS.DynamoDB(dynamoOptions);
}

const generateIndexNames = (indexType, indexNumber) => {
  const indexNames = [];
  for (let i = 0; i < indexNumber; i++) {
    indexNames.push(`${indexType}${i + 1}`);
  }
  return indexNames;
}

const createTable = async (tableName, gsiNumber, lsiNumber, partitionKeyName, sortKeyName, dynamoPort = null) => {
  const dynamoClient = getDynamoClient(dynamoPort);

  const gsiNames = generateIndexNames('GSI', gsiNumber);
  const lsiNames = generateIndexNames('LSI', lsiNumber);

  const creationParams = {
    KeySchema: [
      { AttributeName: partitionKeyName, KeyType: 'HASH' },
      { AttributeName: sortKeyName, KeyType: 'RANGE' }
    ],
    ProvisionedThroughput: {
      ReadCapacityUnits: 5, // TODO make it configurable
      WriteCapacityUnits: 5 // TODO make it configurable
    },
    AttributeDefinitions: [
      { AttributeName: partitionKeyName, AttributeType: 'S' }, // TODO make it configurable
      { AttributeName: sortKeyName, AttributeType: 'S' } // TODO make it configurable
    ],
    TableName: tableName
  };
  if (gsiNames.length > 0) {
    creationParams.GlobalSecondaryIndexes = gsiNames.map(name => ({
      IndexName: name,
      KeySchema: [
        {
          AttributeName: `${name}PK`, // TODO index names should be configurable
          KeyType: 'HASH'
        },
        {
          AttributeName: `${name}SK`,
          KeyType: 'RANGE'
        }
      ],
      Projection: {
        ProjectionType: 'ALL'
      },
      ProvisionedThroughput: {
        ReadCapacityUnits: 5,
        WriteCapacityUnits: 5
      }
    }));
    gsiNames.forEach((name) => {
      creationParams.AttributeDefinitions.push({
        AttributeName: `${name}PK`, AttributeType: 'S'
      });
      creationParams.AttributeDefinitions.push({
        AttributeName: `${name}SK`, AttributeType: 'S'
      });
    });
  }
  if (lsiNames.length) {
    creationParams.LocalSecondaryIndexes = lsiNames.map(name => ({
      IndexName: name,
      KeySchema: [
        {
          AttributeName: partitionKeyName,
          KeyType: 'HASH'
        },
        {
          AttributeName: `${name}SK`,
          KeyType: 'RANGE'
        }
      ],
      Projection: {
        ProjectionType: 'ALL'
      }
    }));
    lsiNames.forEach((name) => {
      creationParams.AttributeDefinitions.push({
        AttributeName: `${name}SK`, AttributeType: 'S'
      });
    });
  }
  await dynamoClient.createTable(creationParams).promise();
};

const listTables = async (dynamoPort = null) => {
  const dynamoClient = getDynamoClient(dynamoPort);
  const { TableNames } = await dynamoClient.listTables({}).promise();
  return TableNames;
};

const deleteTable = async (tableName, dynamoPort) => {
  const dynamoClient = getDynamoClient(dynamoPort);
  const params = {
    TableName: tableName
  };
  await dynamoClient.deleteTable(params).promise();
};

const truncateTable = async (tableName, partitionKeyName, sortKeyName, dynamoPort) => {
  const dynamoClient = getDynamoClient(dynamoPort);
  const deleteItem = (item) => {
    const params = { Key: item, TableName: tableName, ReturnValues: 'ALL_OLD' };
    // TODO could do batch delete instead of 1 by 1
    return dynamoClient.deleteItem(params).promise();
  };
  const { Items } = await scanTable(tableName, partitionKeyName, sortKeyName, dynamoPort);
  const deletePromises = Items.map(i => deleteItem(i));
  await Promise.all(deletePromises);
};

const scanTable = async (tableName, partitionKeyName, sortKeyName, dynamoPort) => {
  const dynamoClient = getDynamoClient(dynamoPort);
  const params = {
    TableName: tableName,
    AttributesToGet: [partitionKeyName, sortKeyName],
    Select: 'SPECIFIC_ATTRIBUTES'
  };
  const itemsRes = [];
  let res = await dynamoClient.scan(params).promise();
  itemsRes.push(...res.Items);
  while (res.LastEvaluatedKey) {
    params.ExclusiveStartKey = res.LastEvaluatedKey;
    res = await dynamoClient.scan(params).promise();
    itemsRes.push(...res.Items);
  }
  return {
    Items: itemsRes
  };
}


module.exports = {
  createTable: createTable,
  deleteTable: deleteTable,
  listTables: listTables,
  scanTable: scanTable,
  truncateTable: truncateTable
};


lib/carpenter.js
const AWS = require('aws-sdk');
const _ = require('lodash');
const UUID = require('uuid');

const generateIndexNames = (indexType, indexNumber) => {
  const indexNames = [];
  for (let i = 0; i < indexNumber; i++) {
    indexNames.push(`${indexType}${i + 1}`);
  }
  return indexNames;
}

class Carpenter {
  constructor(tableName = null, dynamoPort = null) {
    const awsConfig = {
      region: 'local',
      endpoint: 'http://localhost:8000',
      accessKeyId: 'unittest',
      secretAccessKey: 'letmein',
      maxRetries: 5
    };
    if (dynamoPort) {
      awsConfig.endpoint = `http://localhost:${dynamoPort}`;
    }
    AWS.config.update(awsConfig);
    AWS.config.setPromisesDependency();
    this.tableName = tableName || UUID.v4();
    this.dynamoClient = new AWS.DynamoDB({ apiVersion: '2012-08-10' });
  }

  async createTable(partitionKeyName, sortKeyName, gsiNumber = 0, lsiNumber = 0) {
    const gsiNames = generateIndexNames('GSI', gsiNumber);
    const lsiNames = generateIndexNames('LSI', lsiNumber);

    const creationParams = {
      KeySchema: [
        { AttributeName: partitionKeyName, KeyType: 'HASH' },
        { AttributeName: sortKeyName, KeyType: 'RANGE' }
      ],
      ProvisionedThroughput: {
        ReadCapacityUnits: 5, // TODO make it configurable
        WriteCapacityUnits: 5 // TODO make it configurable
      },
      AttributeDefinitions: [
        { AttributeName: partitionKeyName, AttributeType: 'S' }, // TODO make it configurable
        { AttributeName: sortKeyName, AttributeType: 'S' } // TODO make it configurable
      ],
      TableName: this.tableName
    };
    if (gsiNames.length > 0) {
      creationParams.GlobalSecondaryIndexes = gsiNames.map(name => ({
        IndexName: name,
        KeySchema: [
          {
            AttributeName: `${name}PK`, // TODO index names should be configurable
            KeyType: 'HASH'
          },
          {
            AttributeName: `${name}SK`,
            KeyType: 'RANGE'
          }
        ],
        Projection: {
          ProjectionType: 'ALL'
        },
        ProvisionedThroughput: {
          ReadCapacityUnits: 5,
          WriteCapacityUnits: 5
        }
      }));
      gsiNames.forEach((name) => {
        creationParams.AttributeDefinitions.push({
          AttributeName: `${name}PK`, AttributeType: 'S'
        });
        creationParams.AttributeDefinitions.push({
          AttributeName: `${name}SK`, AttributeType: 'S'
        });
      });
    }
    if (lsiNames.length) {
      creationParams.LocalSecondaryIndexes = lsiNames.map(name => ({
        IndexName: name,
        KeySchema: [
          {
            AttributeName: partitionKeyName,
            KeyType: 'HASH'
          },
          {
            AttributeName: `${name}SK`,
            KeyType: 'RANGE'
          }
        ],
        Projection: {
          ProjectionType: 'ALL'
        }
      }));
      lsiNames.forEach((name) => {
        creationParams.AttributeDefinitions.push({
          AttributeName: `${name}SK`, AttributeType: 'S'
        });
      });
    }
    await this.dynamoClient.createTable(creationParams).promise();
  }

  async listTables() {
    const { TableNames } = await this.dynamoClient.listTables({}).promise();
    return TableNames;
  }

  async deleteTable () {
    const params = {
      TableName: this.tableName
    };
    await this.dynamoClient.deleteTable(params).promise();
  }

  async truncateTable(partitionKeyName, sortKeyName) {
    const deleteItem = (item) => {
      const params = { Key: item, TableName: this.tableName, ReturnValues: 'ALL_OLD' };
      // TODO could do batch delete instead of 1 by 1
      return this.dynamoClient.deleteItem(params).promise();
    };
    const { Items } = await this.scanTable(partitionKeyName, sortKeyName);
    const deletePromises = Items.map(i => deleteItem(i));
    await Promise.all(deletePromises);
  }

  async scanTable(partitionKeyName, sortKeyName) {
    const params = {
      TableName: this.tableName,
      AttributesToGet: [partitionKeyName, sortKeyName],
      Select: 'SPECIFIC_ATTRIBUTES'
    };
    const itemsRes = [];
    let res = await this.dynamoClient.scan(params).promise();
    itemsRes.push(...res.Items);
    while (res.LastEvaluatedKey) {
      params.ExclusiveStartKey = res.LastEvaluatedKey;
      res = await this.dynamoClient.scan(params).promise();
      itemsRes.push(...res.Items);
    }
    return {
      Items: itemsRes
    };
  }
}

module.exports = Carpenter;
